### PR TITLE
[MM-30075] Add IsValid method to OpenDialogRequest

### PR DESCRIPTION
**Pull request by JulienTant**
#### Summary
This PR adds an `IsValid() error` method to the following structures: `OpenDialogRequest`, `Dialog` and `DialogElement`, and the associated tests.

#### Ticket Link
https://github.com/mattermost/mattermost-server/issues/16199

#### Release Note
```release-note
Add validation to the open dialog request
```

#### Author comments

I tried to be pretty exhaustive with my tests on the validation, and while the lines of the testcases are pretty long, the test case name usually gives a good indication of what is tested. 

I took some liberty in adding more rules in `DialogElement` - tell me if something has to be changed/removed:
- for a `radio`, at least 2 options are required.
- For `radio` and `select`, default must exist in the in the list of options
- For the name field, the documentation says: `You should use unique name fields in the same dialog`. I interpreted the **should** as a **must**.

I did not mention it in release note as I'm not quite sure about that, but if there was not any validation before, this PR could break modals (too many elements, string too long...)

### CODE REVIEW

File: model/integration_action_test.go [hunk 1]
```diff
@@ -155,3 +156,120 @@ func TestPostActionIntegrationEquals(t *testing.T) {
...
+	makeValidDialogElementArray := func(nb int) (elements []DialogElement) {
+		for i := 0; i < nb; i++ {
+			elements = append(elements, DialogElement{DisplayName: "display", Name: fmt.Sprintf("Name %d", i), Type: "text", SubType: "password", Default: "default", Placeholder: "placeholder", HelpText: "help text", Optional: false, MinLength: 1, MaxLength: 150, DataSource: "", Options: nil})
+		}
+		return elements
+	}
+	invalidDialogElement := DialogElement{}
+
+	testCases := []validableTestCase{
+		{name: "callback id empty", validable: Dialog{CallbackId: "", Title: "title", IntroductionText: "intro text", IconURL: "http://mattermost.com/icon.svg", Elements: makeValidDialogElementArray(1)}, hasError: true},
```
Comment (mickmister) [inline model/integration_action_test.go:195]: If we verify the error has the correct error message, we can assert that this specific validation is the one that failed. @JulienTant What do you think about this? We can have the test name match the error message, and keep the `hasError` boolean.

Also, we can create a helper function to clear up this test data. This makes it much easier to review the tests, including in the future when this code is revisited:

```go
func makeTestDialog(dialogIn Dialog, dependencies... []string) (Dialog, error) {
    dialog := Dialog{
        CallbackId: "callbackid",
        Title: "title",
        IntroductionText: "intro text",
        IconURL: "http://mattermost.com/icon.svg",
        Elements: makeValidDialogElementArray(1),
    }
    
    for _, field := range dependencies {
        switch field {
            case "Title":
                d.Title = dialogIn.Title
            case "CallbackId"
                // ...
            default:
                return dialog, errors.Errorf("unrecognized field name: %s", field)
        }
    }
    
    return dialog, nil
}
```

Then we can use it like so:

```go
testCase := {name: "callback id empty", validable: makeTestDialog(Dialog{CallbackId: ""}, "CallbackId")}
```

File: model/integration_action_test.go [hunk 2]
```diff
@@ -155,3 +156,120 @@ func TestPostActionIntegrationEquals(t *testing.T) {
...
+		s := ""
+		for i := 0; i < length; i++ {
+			s += "a"
+		}
+		return s
+	}
+
+	testCases := []validableTestCase{
+		// general
+		{name: "general: DisplayName empty", validable: DialogElement{DisplayName: "", Name: "name", Type: "text", SubType: "password", Default: "default", Placeholder: "placeholder", HelpText: "help text", Optional: false, MinLength: 0, MaxLength: 3000, DataSource: "", Options: nil}, hasError: true},
```
Comment (mickmister) [inline model/integration_action_test.go:219]: Same suggestion here to use a helper function to clean up the test data duplication here

**Review by mickmister — CHANGES_REQUESTED** [review]
Great work on this @JulienTant! The code looks good :+1: I just have a request to make a test helper to de-duplicate the test data

File: model/integration_action_test.go (hunk 1 above)
Comment (JulienTant) [reply model/integration_action_test.go:195]: Hi @mickmister thanks for the feedback! That's a good idea, let me work on those improvements.

**Comment by JulienTant** [comment]
> Great work on this @JulienTant! The code looks good 👍 I just have a request to make a test helper to de-duplicate the test data

Alright, I pushed changes taking into consideration your feedback. 

The `DialogElement` part made the initial idea of using test name as error message a bit challenging, as we have a lot of similar test but on different types. So I introduced an optionnal `ErrorMsg` field so that we can keep descriptive names with specific error messages when needed.

**Review by mickmister — APPROVED** [review]
Thanks for cleaning up the test data @j8r! LGTM :+1:

**Comment by mickmister** [comment]
@DHaussermann Manual testing on this PR would be quite tedious. I recommend looking over lines 304-384 in the changes for `model/integration_action_test.go` and gauge how you feel from there. THere seems to be a test case for each of the requirements laid out in the docs.

**Comment by emilyacook** [comment]
Thanks for participating in Hacktoberfest! You can claim your sticker set here: https://get.printfection.com/hacktober21/4144583267 @JulienTant

**Comment by JulienTant** [comment]
> @DHaussermann Manual testing on this PR would be quite tedious. I recommend looking over lines 304-384 in the changes for `model/integration_action_test.go` and gauge how you feel from there. THere seems to be a test case for each of the requirements laid out in the docs.

Hi Mattermost team,

I decided to test this with the playbook plugin and realized it does not work because the plugin use relative path rather than absolute URL in the OpenDialogRequest. Also the plugin demo promote both: https://github.com/mattermost/mattermost-plugin-demo/blob/master/server/command_hooks.go

Should I drop this verification completely? Or will we end up enforcing absolute URL?

**Comment by JulienTant** [comment]
It's done!

**Review by mickmister — APPROVED** [review]
LGTM :tada: Thanks @JulienTant!

**Review by larkox — APPROVED** [review]
LGTM

**Comment by jasonblais** [comment]
@DHaussermann friendly reminder to help with a QA review :)

**Comment by JulienTant** [comment]
Do we still need this?! 😅

**Comment by hanzei** [comment]
@JulienTant I still would like to get this PR in

**Comment by JulienTant** [comment]
Yay the PR is up to date :D

**Comment by JulienTant** [comment]
Hi @DHaussermann - Is there anything I can do to help you review this?

**Pull request closed without merging**